README.md

# smetlite

<!-- badges: start -->
[![R-CMD-check](https://github.com/baseldatascience/smetlite/actions/workflows/R-CMD-check.yaml/badge.svg)](https://github.com/baseldatascience/smetlite/actions/workflows/R-CMD-check.yaml)
<!-- badges: end -->

The goal of smetlite is to ease working with SMET files from meteorology.
The specification of these files you find in
https://models.slf.ch/docserver/meteoio/SMET_specifications.pdf.

A few short example smet files are included in the package;
real life examples you can for instance download from 
https://www.envidat.ch/dataset/38112bee-f865-4b29-ba19-47a99ba78833/resource/3074d1dc-6083-4025-9ac4-484cfa88658b/download/wfj_optimaldataset_v7.smet
or https://www.envidat.ch/dataset/a9264e9f-ea9b-4c89-b471-464efb891826/resource/a5918c50-6716-4dcf-b462-e1970ae1e072/download/wfj.smet.

Suggestions for useful extension of this tiny package are very welcome!

## Installation

You can install the development version of smetlite from [GitHub](https://github.com/) with:

``` r
# install.packages("devtools")
devtools::install_github("baseldatascience/smetlite")
```

## Example

Simple example how to read in a smet file:

``` r
library(smetlite)
read_smet(system.file('examples/test.smet', package = 'smetlite'))
```


NEWS.md
# smetlite 0.2.10

* migrate to BaselDataScience repo

# smetlite 0.2.7

* Initial CRAN submission.


R/test_smet.R
#' Test SMET data
#'
#' Artificial data derived from rifugio Vaccarone
#'
#' @format ## `test_smet`
#' A data frame with 47 rows and 5 columns:
#' \describe{
#'   \item{timestamp}{Timestamp of measurement}
#'   \item{VW}{Velocity Wind, in m/s}
#'   \item{DW}{Direction Wind, in degrees, clockwise and north being zero degrees}
#'   \item{TA}{Temperature Air, in Kelvin}
#'   \item{HS}{Height Snow, in m}
#' }
#' @source <http://webgis.arpa.piemonte.it/webmeteo/meteo.php?CODTOT=001114901>
"test_smet"


tests/tinytest.R
if ( requireNamespace("tinytest", quietly=TRUE) ){
  tinytest::test_package("smetlite")
}


inst/tinytest/test_smetlite.R
# Check with simple test SMET file
xx <- read_smet(system.file('extdata', 'test.smet', package = 'smetlite'))
attr(xx, 'file') <- 'test.smet'
expect_equal(xx, test_smet)


R/zzz.R
#' A named list explaining the permissible time and measurement variable names.
#' @format ## `smet_dict`
#' \describe{
#'   \item{P}{Air pressure, in Pa}
#'   \item{TA}{Temperature Air, in Kelvin}
#'   \item{TSS}{Temperature Snow Surface, in Kelvin}
#'   \item{TSG}{Temperature Surface Ground, in Kelvin}
#'   \item{RH}{Relative Humidity, between 0 and 1}
#'   \item{VW_MAX}{Maximum wind velocity, in m/s}
#'   \item{VW}{Velocity Wind, in m/s}
#'   \item{DW}{Direction Wind, in degrees, clockwise and north being zero degrees}
#'   \item{ISWR}{Incoming Short Wave Radiation, in W/m2}
#'   \item{OSWR}{Reflected Short Wave Radiation, in W/m2}
#'   \item{ILWR}{Incoming Long Wave Radiation, in W/m2}
#'   \item{OLWR}{Outgoing Long Wave Radiation, in W/m2}
#'   \item{PINT}{Precipitation Intensity, in mm/h, as an average over the timestep}
#'   \item{PSUM}{Precipitation accumulation, in mm, summed over the last timestep}
#'   \item{HS}{Height Snow, in m}
#'   \item{timestamp}{ISO 8601 Combined date and time formatted}
#'   \item{julian}{the decimal number of days and fractions of a day since January 1, 4713 BC Greenwich noon, Julian proleptic calendar3. If both timestamps and julian are present, they must be within less than 1 second of each other.}
#' }
#' @source https://models.slf.ch/docserver/meteoio/SMET_specifications.pdf
#' @export
smet_dict <- list(
  P="Air pressure, in Pa",
  TA="Temperature Air, in Kelvin",
  TSS="Temperature Snow Surface, in Kelvin",
  TSG="Temperature Surface Ground, in Kelvin",
  RH="Relative Humidity, between 0 and 1",
  VW_MAX="Maximum wind velocity, in m/s",
  VW="Velocity Wind, in m/s DW Direction Wind, in degrees, clockwise and north being zero degrees",
  ISWR="Incoming Short Wave Radiation, in W/m2",
  OSWR="Reflected Short Wave Radiation, in W/m2",
  ILWR="Incoming Long Wave Radiation, in W/m2",
  OLWR="Outgoing Long Wave Radiation, in W/m2",
  PINT="Precipitation Intensity, in mm/h, as an average over the timestep",
  PSUM="Precipitation accumulation, in mm, summed over the last timestep",
  HS="Height Snow, in m",
  timestamp="ISO 8601 Combined date and time formatted",
  julian="as the decimal number of days and fractions of a day since January 1, 4713 BC Greenwich noon, Julian proleptic calendar3. If both timestamps and julian are present, they must be within less than 1 second of each other."
)


R/legend.R
# list of permissible tiem and measurement variables
legend <- list(
  P="Air pressure, in Pa",
  TA="Temperature Air, in Kelvin",
  TSS="Temperature Snow Surface, in Kelvin",
  TSG="Temperature Surface Ground, in Kelvin",
  RH="Relative Humidity, between 0 and 1",
  VW_MAX="Maximum wind velocity, in m/s",
  VW="Velocity Wind, in m/s DW Direction Wind, in degrees, clockwise and north being zero degrees",
  ISWR="Incoming Short Wave Radiation, in W/m2",
  OSWR="Reflected Short Wave Radiation, in W/m2",
  ILWR="Incoming Long Wave Radiation, in W/m2",
  OLWR="Outgoing Long Wave Radiation, in W/m2",
  PINT="Precipitation Intensity, in mm/h, as an average over the timestep",
  PSUM="Precipitation accumulation, in mm, summed over the last timestep",
  HS="Height Snow, in m",
  timestamp="ISO 8601 Combined date and time formatted",
  julian="as the decimal number of days and fractions of a day since January 1, 4713 BC Greenwich noon, Julian proleptic calendar3. If both timestamps and julian are present, they must be within less than 1 second of each other."
)


R/smet-functions.R
# functions around smet class

read_smet <- function(filename) {
  tmp <- readLines(filename, n = 999, warn = FALSE,
              encoding = 'UTF-8')
  smet_version <- tmp[1]

  # determine line number of section headers
  i_data <- grep('^\\[DATA\\]', tmp)
  i_header <- grep('^\\[HEADER\\]', tmp)
  # work on smet header
  tmp <- tmp[(i_header+1):(i_data-1)]
  tmp <- sub('[;#].*', '', tmp)
  tmp <- tmp[grepl('=', tmp, fixed = TRUE)]

  xx <- stringr::str_split_fixed(tmp, '=', n = 2)
  yy <- structure(sub('\\s', '', xx[,2]), names=stringr::str_squish(xx[,1]))
  nm <- unique(names(yy))
  header <- structure(lapply(lapply(nm, function(X) names(yy)==X),
                             function(X) paste(yy[X], collapse = '\n')),
                      names=nm)

  # remove leading and trailing whitespace and uniquify repeated whitespace within
  for (x in setdiff(nm, 'note')) header[[x]] <- stringr::str_squish(header[[x]])

  # vectorise
  for (x in intersect(c('fields', 'units_multiplier', 'units_offset'), nm))
    header[x] <- stringr::str_split(header[x], '\\s', n=Inf)

  # convert numeric coordinates
  for (x in intersect(c('altitude', 'easting', 'latitude', 'longitude', 'northing',
                        'tz', 'units_multiplier', 'units_offset'), nm))
    header[[x]] <- as.numeric(header[[x]])
  if (!is.null(header$epsg)) header$epsg <- as.integer(header$epsg)


  # read the data
  col_types <- structure(rep(list(readr::col_double()), each=length(header$fields)),
                         names = header$fields)
  col_types$timestamp <- readr::col_datetime()
  dat <- readr::read_table(filename,
                           skip = i_data,
                           col_names = header$fields,
                           col_types = col_types,
                           locale = readr::locale(tz='UTC'),
                           na = header$nodata)

  # convert timestamp to UTC timezone
  if (!is.null(header$tz)) dat$timestamp <- dat$timestamp - header$tz * 3600

  # back to MKSA units
  for (i in seq_along(header$units_multiplier)) {
    if (header$units_multiplier[i] != '1') dat[, i] <- as.numeric(header$units_multiplier[i]) * dat[,i]
  }
  for (i in seq_along(header$units_offset)) {
    if (header$units_offset[i] != '0') dat[, i] <- as.numeric(header$units_offset[i]) + dat[,i]
  }
  methods::new('smet', signature=smet_version, header=header, data=dat, file=filename)
}


R/smet.R
# functions around smet class

#' Read a SMET file into a smet dataframe
#'
#' @param filename file or connection to be read by readr::read_table()
#'
#' @return dataframe with all
#' @export
#'
#' @examples read_smet(system.file('examples/test.smet', package = 'smetlite'))
read_smet <- function(filename) {
  tmp <- readLines(filename, n = 999, warn = FALSE,
                   encoding = 'UTF-8')
  smet_version <- tmp[1]

  # determine line number of section headers
  i_data <- grep('^\\[DATA\\]', tmp)
  i_header <- grep('^\\[HEADER\\]', tmp)
  # work on smet header
  tmp <- tmp[(i_header+1):(i_data-1)]
  tmp <- sub('[;#].*', '', tmp)
  tmp <- tmp[grepl('=', tmp, fixed = TRUE)]

  xx <- stringr::str_split_fixed(tmp, '=', n = 2)
  yy <- structure(sub('\\s', '', xx[,2]), names=stringr::str_squish(xx[,1]))
  nm <- unique(names(yy))
  header <- structure(lapply(lapply(nm, function(X) names(yy)==X),
                             function(X) paste(yy[X], collapse = '\n')),
                      names=nm)

  # remove leading and trailing whitespace and uniquify repeated whitespace within
  for (x in setdiff(nm, 'note')) header[[x]] <- stringr::str_squish(header[[x]])

  # vectorise
  for (x in intersect(c('fields', 'units_multiplier', 'units_offset'), nm))
    header[x] <- stringr::str_split(header[x], '\\s', n=Inf)

  # convert numeric coordinates
  for (x in intersect(c('altitude', 'easting', 'latitude', 'longitude', 'northing',
                        'tz', 'units_multiplier', 'units_offset', 'nodata'), nm))
    header[[x]] <- vapply(str2expression(header[[x]]), eval, double(1))
  if (!is.null(header$epsg)) header$epsg <- as.integer(header$epsg)


  # read the data
  col_types <- structure(rep(list(readr::col_double()), each=length(header$fields)),
                         names = header$fields)
  col_types$timestamp <- readr::col_datetime()
  dat <- readr::read_table(filename,
                           skip = i_data,
                           col_names = header$fields,
                           col_types = col_types,
                           locale = readr::locale(tz='UTC')
                           )

  # convert timestamp to UTC timezone
  if (!is.null(header$tz)) dat$timestamp <- dat$timestamp - header$tz * 3600

  # make sure nodata is interpreted as NA
  dat[dat==header$nodata] <- NA

  # back to MKSA units
  adj_mult <- header$units_multiplier != 1
  dat[, adj_mult] <- dat[, adj_mult] * rep(header$units_multiplier[adj_mult], each=nrow(dat))
  adj_add <- header$units_offset != 0
  dat[, adj_add] <- dat[, adj_add] + rep(header$units_offset[adj_add], each=nrow(dat))

  # attach metadata as attributes
  attr(dat, 'signature') <- smet_version
  attr(dat, 'header') <- header
  attr(dat, 'file') <- filename
  attr(dat, 'class') <- c('smet', attr(dat, 'class', exact = TRUE))
  dat
}

#' retrieve header
#'
#' @param smet object of class smet
#'
#' @return named list of header information
#' @export
#'
#' @examples header(read_smet(system.file('examples/test.smet', package = 'smetlite')))
header <- function(smet) {
  attr(smet, 'header')
}
